feat: new UI, upgrade bulma 1.0.3

candidosales opened the pull request:
**Why**

When I first came into contact with the tool, I found the style a bit intimidating, and there were opportunities for improvement, so I'm creating this PR to improve the developer's life.

**What**

- [x] Upgrade [Bulma 1.0.3](https://github.com/jgthms/bulma/releases/tag/1.0.3)

## New home

Reduce the weight of the typography and improve the colours

![CleanShot 2025-02-08 at 07 29 36@2x](https://github.com/user-attachments/assets/2c7be651-36d6-4bf6-b350-d2e0d0012341)

## New task UI

Splitting the inputs and arguments sections into three columns will prevent the user from scrolling through the page.
 
![CleanShot 2025-02-08 at 08 22 06@2x](https://github.com/user-attachments/assets/a39cad60-01a0-4c2f-995d-befeb7eb2993)

## Saving spacing

I added accordion sections to the Source code and Runs. The first run section will always be open.

https://github.com/user-attachments/assets/ed45b2f4-5978-41c1-829b-e3b4b8d34c5a


Please feel free to share any feedback - I'd love to hear your perspective! 😊

candidosales commented:
@adrianna-chang-shopify, @etiennebarrie  and @andrewn617 , could you review it 🙏?

nvasilevski commented on test/dummy/db/schema.rb:
```diff
@@ -10,13 +10,13 @@
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2023_06_22_035229) do
+ActiveRecord::Schema[7.2].define(version: 2023_06_22_035229) do
```
This was most likely committed unintentionally, let's revert it

nvasilevski reviewed (commented):
Thanks for the suggestion! Could you squash your commits into one and what's more importantly remove the irrelevant schema changes? I assume it was added unintentionally. Thanks!

candidosales commented:
@adrianna-chang-shopify , thanks for the feedback! I'll do it in soon

adrianna-chang-shopify commented:
Can you squash into fewer commits please and then I'll take a look? Thanks!

candidosales commented:
@nvasilevski, could you review it? 🙏🏼

nvasilevski commented on app/helpers/maintenance_tasks/tasks_helper.rb:
```diff
@@ -149,26 +149,38 @@ def resolve_inclusion_value(task, parameter_name)
 
+    # Return whether the parameter should be rendered as a dropdown list.
+    def select?(form_builder, parameter_name)
+      inclusion_values = resolve_inclusion_value(form_builder.object, parameter_name)
+      inclusion_values.is_a?(Array)
```
nit but I think it's going to be even more readable (and shorter) if inlined
```suggestion
      resolve_inclusion_value(form_builder.object, parameter_name).is_a?(Array)
```

nvasilevski commented:
>@nvasilevski, could you review it? 🙏🏼

Correct me if I'm wrong but I think @adrianna-chang-shopify suggested we split this PR even further. One will be a dependabot-like PR that just upgrades the library and makes no changes to the UI (unless required by changes in the library). This one has a clear benefit and can be merged without discussion

But changes to the home page layout and changes to the attributes layout deserve separate PRs. Sorry for the troubles but making PRs that focus on one specific thing makes it much easier to review, to make a decision to accept a PR and allows us to revert only one part of the proposal if we suddenly find out that it wasn't completely suitable for all cases. Appreciate your cooperation!

adrianna-chang-shopify commented on app/helpers/maintenance_tasks/application_helper.rb:
```diff
@@ -18,5 +18,12 @@ def time_ago(datetime)
       end
     end
+
+    def attribute_required?(task_data_show, attribute_name)
```
Should this be in the `tasks_helper` instead?

candidosales replied:
Removed, I'll take this in the following PR

adrianna-chang-shopify reviewed (commented):
There are a lot of whitespace / indentation changes in this PR that make it difficult to review. Can you ensure the spacing is cleaned up please?

Additionally, we'd ideally limit the number of code changes in this PR. I understand needing to make a few changes to HTML classes etc. in this PR, given that Bulma changed some of their styles, as you mentioned. But it'll be easier to get this reviewed + merged if we avoid bundling in "new" changes with this PR - things like helpers, changes to what's passed to partials, adding the run id to `_serializable`, etc. Can we do those in a follow up?

Thanks for your understanding!

candidosales commented on app/views/maintenance_tasks/runs/_run.html.erb:
```diff
@@ -1,9 +1,19 @@
-    <%= status_tag run.status %>
-    <span class="is-pulled-right" title="Run ID">#<%= run.id %></span>
-  </h5>
+<% if run_counter == 0 %>
```
the counter is available in collection partials; you can see more here: https://www.mintbit.com/blog/rails-counter-available-in-collection-partials/

candidosales commented on app/views/maintenance_tasks/tasks/show.html.erb:
```diff
@@ -1,41 +1,48 @@
+			  <% required_class = attribute_required?(@task, parameter_name) ? 'is-required' : '' %>
+              <%= ff.label parameter_name, parameter_name, class: "label is-family-monospace  #{required_class}" %>
+				<div class="<%= select?(ff, parameter_name) ? "select" : "control" %>">
+					<%= parameter_field(ff, parameter_name) %>
```
Great! I have refactored.

candidosales commented:
@adrianna-chang-shopify , thanks for the feedback. I have addressed your suggestions. Other improvements I'll take in the following PRs.

It's ready to review it.

nvasilevski commented on app/views/maintenance_tasks/runs/_serializable.html.erb:
```diff
@@ -1,24 +1,21 @@
+          <code><%= value %></code>
+        <% end %>
+      <% end %>
+    </div>
+    </div>
```
It looks like `div` on line 16 should be nested into the outer `div`

candidosales commented:
@adrianna-chang-shopify , could you review it? 🙏🏼

adrianna-chang-shopify reviewed (commented):
This is a lot easier to review, thanks!

Posting screenshots here for comparison

**Previous UI**
![Screenshot 2025-02-21 at 1 50 41 PM](https://github.com/user-attachments/assets/588d5b7b-192f-4bd7-9e6e-8fdcb3d3849a)

**New UI**
![Screenshot 2025-02-21 at 1 50 12 PM](https://github.com/user-attachments/assets/07fa55e5-c98e-4e7f-941a-ffa0ca6ce10a)

I preferred the task name on the show page when it was larger -- can we make the title a bit larger there? Something in between the two maybe?

I'm also finding the status labels harder to read / differentiate, can we keep the "regular" styling rather than the light version?

candidosales commented:
@adrianna-chang-shopify ,

**I have increased the font size to 1.5rem (`is-4`) to 2rem (`is-3`)**

![CleanShot 2025-02-21 at 17 09 41@2x](https://github.com/user-attachments/assets/37108e20-b448-4645-96e7-e6489939a827)

**Removed the light version in status**

### Before

![CleanShot 2025-02-21 at 17 13 43@2x](https://github.com/user-attachments/assets/84170f03-8b92-49f6-a536-0e0379f52753)


### After

![CleanShot 2025-02-21 at 17 17 26@2x](https://github.com/user-attachments/assets/3422be9a-761a-40e2-b97a-52dabc153b7b)

adrianna-chang-shopify commented on app/views/maintenance_tasks/runs/_run.html.erb:
```diff
@@ -1,9 +1,15 @@
-    <%= status_tag run.status %>
-    <span class="is-pulled-right" title="Run ID">#<%= run.id %></span>
-  </h5>
+<%= tag.details(class: "box", open: run_counter == 0 ? true : false) do %>
```
Can we change this so that, rather than showing just the first run in the collection, it also shows _all_ active runs? IMO if you have multiple active runs happening at the same time it would be nice if the details showed for all of them.

candidosales replied:
Yeah, great! I'll do it

adrianna-chang-shopify commented on app/views/layouts/maintenance_tasks/application.html.erb:
```diff
@@ -29,6 +29,31 @@
       .ruby-kw { color: #d73a49; }
       .ruby-label, .ruby-tstring-beg, .ruby-tstring-content, .ruby-tstring-end { color: #032f62; }
+
+      .select, select { width: 100% }
+      summary { cursor: pointer; }
+      input[type="datetime-local"], input[type="date"], input[type="time"] {
+      	width: fit-content
+      }
+	  	details>summary {
+      	list-style: none;
+      }
+      summary::-webkit-details-marker {
+      	display: none
+      }
+      summary::before {
+      	content: '► ';
+      	position:absolute;
+      	font-size: 16px
+      }
+      details[open] summary:before {
+      	content: "▼ ";
+      }
+
+      .box {
+      	--tw-shadow: 0 4px 6px -1px  #0000001a, 0 2px 4px -2px  #0000001a;
+          box-shadow: 0 0 #0000, 0 0 #0000,  0 0 #0000, 0 0 #0000, var(--tw-shadow);
+      }
```
Here's the CSP hash: `'sha256-D5WBghavYzDNLckWJrp0/Lb4V/AcUC9TmavYVFuqdIQ='`

candidosales commented:
@adrianna-chang-shopify , could you review it? 🙏🏼

adrianna-chang-shopify commented on app/views/layouts/maintenance_tasks/application.html.erb:
```diff
@@ -29,6 +29,31 @@
+      input[type="datetime-local"], input[type="date"], input[type="time"] {
+        width: fit-content;
+      }
+	  details>summary {
```
Something's still off with the spacing here.

adrianna-chang-shopify commented on app/views/maintenance_tasks/runs/_serializable.html.erb:
```diff
@@ -1,24 +1,21 @@
+            <% if !value.empty? %>
+              <% if value.include?("\n") %>
+                <pre><%= value %></pre>
+        	  <% else %>
```
```suggestion
              <% else %>
```

candidosales replied:
This one is correct as I mentioned here: https://github.com/Shopify/maintenance_tasks/pull/1163#discussion_r1970171191

adrianna-chang-shopify commented on app/views/maintenance_tasks/tasks/index.html.erb:
```diff
@@ -9,15 +9,17 @@
-      <%= render partial: 'task', collection: new_tasks %>
+      <h3 class="title is-4 has-text-weight-bold">New Tasks</h3>
+	  <div class="grid is-col-min-20">
+      	<%= render partial: 'task', collection: new_tasks %>
```
```suggestion
      	    <%= render partial: 'task', collection: new_tasks %>
```

adrianna-chang-shopify reviewed (commented):
Couple more spacing / indentation nits, but otherwise this is looking good!

Can you please squash into a single commit with a meaningful message? You can follow the guidelines [here](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#commit-your-changes).

candidosales commented:
@adrianna-chang-shopify , I have updated with `npm/bulma@1.0.3/css/versions/bulma-no-dark-mode.min.css`. Could you review it?

candidosales commented:
These failures are due to unreliable tests; could you please rerun them?

candidosales commented:
@adrianna-chang-shopify , I did some investigation and `rails test:system` is pretty unstable (`test/system/maintenance_tasks/runs_test.rb`).

I have run many times and received various test errors. I think the root cause is related to a race condition and how the jobs ran so quickly. I have tried in the main branch, and these issues are occurring as well, so I believe it's a known issue.

![CleanShot 2025-03-14 at 08 46 43@2x](https://github.com/user-attachments/assets/b1ad5bae-2a24-4b0e-a4b0-7d0461f7a33e)

![CleanShot 2025-03-14 at 08 46 31@2x](https://github.com/user-attachments/assets/7d51a4bc-27f6-4c35-a790-afa055beaf74)

nvasilevski reviewed (approved):
Yeah, unfortunately system tests haven't been very stable recently. But I think this PR is ready to be merged

adrianna-chang-shopify commented:
Arf, yeah we really need to overhaul the system tests 😞 @etiennebarrie I believe started looking into [playwright as an alternative](https://github.com/Shopify/maintenance_tasks/issues/1176#issuecomment-2721713568).

Tests are also failing on main so I'll ship this. Thanks for your work! 🙇‍♀️
